File: test-suite-java/src/test/java/io/micronaut/el/test/CatalogExpressionsTest.java
package io.micronaut.el.test;

import io.micronaut.el.CompiledELContext;
import jakarta.el.LambdaExpression;
import jakarta.el.ValueExpression;
import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertEquals;

class CatalogExpressionsTest {

    private final List<Book> books = List.of(
        new Book("Expression Language", "history", 20d),
        new Book("Sourcegen", "computing", 5d)
    );
    private final Author author = new Author("Denis", books);
    private final CompiledELContext context = new CompiledELContext()
        .setBean("author", author)
        .setBean("books", books)
        .setBean("strings", new String[]{"a", "b"})
        .setBean("customer", author);

    @Test
    void staticallyTypedResolution() {
        assertEquals("Denis", value(CatalogExpressions$ELExpressions.AUTHOR_NAME));
        assertEquals("Hi, Denis", value(CatalogExpressions$ELExpressions.GREETING));
        assertEquals("Expression Language", value(CatalogExpressions$ELExpressions.FIRST_TITLE));
    }

    @Test
    void dynamicResolutionUsesTheGeneratedResolvers() {
        assertEquals("Denis", value(CatalogExpressions$ELExpressions.DYNAMIC_NAME));
        assertEquals("history", value(CatalogExpressions$ELExpressions.DYNAMIC_CATEGORY));
    }

    @Test
    void functions() {
        assertEquals(5, value(CatalogExpressions$ELExpressions.NAME_LENGTH));
        assertEquals("DENIS", value(CatalogExpressions$ELExpressions.UPPER_NAME));
        assertEquals("a,b", value(CatalogExpressions$ELExpressions.FUNCTION_JOIN));
        assertEquals("a,b", value(CatalogExpressions$ELExpressions.FUNCTION_ARRAY_JOIN));
        assertEquals("a,b", value(CatalogExpressions$ELExpressions.STATIC_JOIN));
    }

    @Test
    void staticReferencesAndConstructors() {
        assertEquals(Suit.SPADE, value(CatalogExpressions$ELExpressions.SUIT));
        assertEquals(Suit.SPADE, value(CatalogExpressions$ELExpressions.IMPORTED_SUIT));
        assertEquals(Boolean.TRUE, value(CatalogExpressions$ELExpressions.BOOLEAN_CONSTANT));
        assertEquals(42, value(CatalogExpressions$ELExpressions.STATIC_METHOD));
        assertEquals("EL", ((Book) value(CatalogExpressions$ELExpressions.NEW_BOOK)).getTitle());
        assertEquals("a,b", value(CatalogExpressions$ELExpressions.VARARGS_CONSTRUCTOR));
        assertEquals("a,b", value(CatalogExpressions$ELExpressions.VARARGS));
    }

    @Test
    void collectionOperations() {
        assertEquals(List.of("Expression Language"), value(CatalogExpressions$ELExpressions.EXPENSIVE));
        assertEquals(25d, value(CatalogExpressions$ELExpressions.TOTAL));
        assertEquals("Expression Language", value(CatalogExpressions$ELExpressions.MOST_EXPENSIVE));
    }

    @Test
    void lambdaExpressions() {
        assertEquals(7L, value(CatalogExpressions$ELExpressions.IMMEDIATE_LAMBDA));
        assertEquals(7L, value(CatalogExpressions$ELExpressions.ASSIGNED_LAMBDA));
        assertEquals(120L, value(CatalogExpressions$ELExpressions.FACTORIAL));
        LambdaExpression outer = (LambdaExpression) value(CatalogExpressions$ELExpressions.NESTED_LAMBDA);
        LambdaExpression inner = (LambdaExpression) outer.invoke(context, 3L);
        assertEquals(7L, inner.invoke(context, 4L));
    }

    private Object value(ValueExpression expression) {
        return expression.getValue(context);
    }
}

File: doc-examples/example-kotlin/src/test/kotlin/example/CatalogExpressionsTest.kt
package example

import io.micronaut.el.CompiledELContext
import jakarta.el.ELManager
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Test

class CatalogExpressionsTest {

    @Test
    fun functionsImportsAndCollections() {
        val context = CompiledELContext()
            .setBean("book", Book("Jakarta EL", "reference", 20.0))
            .setBean("books", listOf(Book("Jakarta EL", "reference", 20.0), Book("Leaflet", "marketing", 2.0)))
        val factory = ELManager.getExpressionFactory()

        assertEquals("JAKARTA EL!", factory.createValueExpression(context, "\${text:shout(book.title)}", String::class.java).getValue(context))
        assertEquals("JE", factory.createValueExpression(context, "\${text:initials(book.title)}", String::class.java).getValue(context))
        assertEquals(25.0, factory.createValueExpression(context, "\${Math.max(book.unitPrice, 25.0)}", Double::class.java).getValue(context))
        assertEquals(listOf("Jakarta EL"), factory.createValueExpression(context,
            "\${books.stream().filter(b -> b.unitPrice > 10).map(b -> b.title).toList()}", List::class.java).getValue(context))
        assertEquals(40.0, factory.createValueExpression(context, "\${(price -> price * 2)(book.unitPrice)}", Double::class.java).getValue(context))
        assertEquals(15.0, factory.createValueExpression(context,
            "\${discount = (price, percent) -> price * (100 - percent) / 100; discount(book.unitPrice, 25)}", Double::class.java).getValue(context))
        assertEquals(listOf("Leaflet", "Jakarta EL"), factory.createValueExpression(context,
            "\${books.stream().sorted((a, b) -> a.unitPrice - b.unitPrice).map(b -> b.title).toList()}", List::class.java).getValue(context))
    }
}
